pkg/koordlet/resmanager/cpu_evict.go

```diff
@@ -0,0 +1,362 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resmanager
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	slov1alpha1 "github.com/koordinator-sh/koordinator/apis/slo/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/features"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+const (
+	beCPUSatisfactionLowPercentMax   = 60
+	beCPUSatisfactionUpperPercentMax = 100
+	beCPUUsageHighEnough             = 0.9
+)
+
+type CPUEvictor struct {
+	resmanager    *resmanager
+	lastEvictTime time.Time
+}
+
+func NewCPUEvictor(resmanager *resmanager) *CPUEvictor {
+	return &CPUEvictor{
+		resmanager:    resmanager,
+		lastEvictTime: time.Now(),
+	}
+}
+
+type podEvictCPUInfo struct {
+	milliRequest   int64
+	milliUsedCores int64
+	// cpuUsage = milliUsedCores / milliRequest
+	cpuUsage float64
+	pod      *corev1.Pod
+}
+
+func (c *CPUEvictor) cpuEvict() {
+	klog.Info("cpu evict process start")
+
+	if time.Now().Before(c.lastEvictTime.Add(time.Duration(c.resmanager.config.CPUEvictCoolTimeSeconds) * time.Second)) {
+		klog.V(5).Infof("skip CPU evict process, still in evict cool time")
+		return
+	}
+
+	nodeSLO := c.resmanager.getNodeSLOCopy()
+	if disabled, err := isFeatureDisabled(nodeSLO, features.BECPUEvict); err != nil {
+		klog.Warningf("cpuEvict failed, cannot check the feature gate, err: %s", err)
+		return
+	} else if disabled {
+		klog.V(5).Infof("cpuEvict skipped, nodeSLO disable the feature gate")
+		return
+	}
+
+	thresholdConfig := nodeSLO.Spec.ResourceUsedThresholdWithBE
+	windowSeconds := c.resmanager.collectResUsedIntervalSeconds * 2
+	if thresholdConfig.CPUEvictTimeWindowSeconds != nil && *thresholdConfig.CPUEvictTimeWindowSeconds > c.resmanager.collectResUsedIntervalSeconds {
+		windowSeconds = *thresholdConfig.CPUEvictTimeWindowSeconds
+	}
+
+	node := c.resmanager.statesInformer.GetNode()
+	if node == nil {
+		klog.Warningf("cpuEvict failed, got nil node %s", c.resmanager.nodeName)
+		return
+	}
+
+	cpuCapacity := node.Status.Capacity.Cpu().Value()
+	if cpuCapacity <= 0 {
+		klog.Warningf("cpuEvict failed, node cpuCapacity not valid,value: %d", cpuCapacity)
+		return
+	}
+
+	isNeedEvictAll, message := c.isCritical(node, thresholdConfig, windowSeconds)
+	if isNeedEvictAll {
+		c.killAndEvictAllBEPods(node, message)
```

Comment:
If CPU usage of node is over the threshold, is it crude to evict all BE pods ? I think we need a dynamic evicting speed based on real-time  CPUUsage - threshold.

Reply:
actually only the rarest condition would trigger this function. I'll try to optimize this in the next iteration.